File: android/app/src/main/java/com/commutescout/drive/Report.kt
package com.commutescout.drive

import android.app.Activity
import androidx.compose.foundation.border
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AcUnit
import androidx.compose.material.icons.filled.Block
import androidx.compose.material.icons.filled.CameraAlt
import androidx.compose.material.icons.filled.Cloud
import androidx.compose.material.icons.filled.DirectionsCar
import androidx.compose.material.icons.filled.LocalPolice
import androidx.compose.material.icons.filled.Map
import androidx.compose.material.icons.filled.Send
import androidx.compose.material.icons.filled.Traffic
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material.icons.filled.Water
import androidx.compose.material.icons.filled.Link
import androidx.compose.material.icons.filled.LinkOff
import androidx.compose.material.icons.filled.CarCrash
import androidx.compose.material.icons.filled.RemoveRoad
import androidx.compose.material3.Button
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/** What a driver can report, the website's list. Kinds are the Flare vocabulary. */
object ReportKinds {
    data class Kind(val kind: String, val label: String, val icon: ImageVector)
    val all = listOf(
        Kind("POLICE_VISIBLE", "Police", Icons.Default.LocalPolice),
        Kind("CRASH_MAJOR", "Crash", Icons.Default.CarCrash),
        Kind("HAZARD_ON_ROAD", "Hazard on road", Icons.Default.Warning),
        Kind("HAZARD_SHOULDER_CAR", "Car on shoulder", Icons.Default.DirectionsCar),
        Kind("ROAD_CLOSED", "Road closed", Icons.Default.Block),
        Kind("LANE_CLOSED", "Lane closed", Icons.Default.RemoveRoad),
        Kind("JAM_HEAVY", "Traffic jam", Icons.Default.Traffic),
        Kind("WEATHER_FOG", "Fog or weather", Icons.Default.Cloud),
        Kind("WEATHER_FLOOD", "Flooding", Icons.Default.Water),
        Kind("WEATHER_ICE", "Ice or snow", Icons.Default.AcUnit),
        Kind("CHAINS_REQUIRED", "Chains required", Icons.Default.Link),
        Kind("CHAINS_NOT_REQUIRED", "Chains not needed", Icons.Default.LinkOff),
        Kind("CAMERA_ISSUE", "Camera issue", Icons.Default.CameraAlt),
        Kind("MAP_ISSUE", "Map issue", Icons.Default.Map),
    )
}

/** Sends a report to commutescout.com, which fans it out to plugins. */
object Reporter {
    suspend fun send(kind: String, lat: Double, lon: Double, heading: Double?, description: String, token: String) {
        val body = buildString {
            append("""{"kind":"$kind","lat":$lat,"lon":$lon,"client":"commutescout-android/${BuildConfig.VERSION_NAME}"""")
            if (heading != null && heading >= 0) append(""","heading_deg":$heading""")
            if (description.isNotBlank()) append(""","description":${JsonPrimitive(description)}""")
            append("}")
        }
        val (status, text) = Backend.send("POST", "/api/flare/report", token, body)
        if (status != 201 && status != 202) {
            val msg = runCatching { Backend.json.parseToJsonElement(text).jsonObject["error"] }.getOrNull()
            val detail = (msg as? JsonPrimitive)?.content ?: (msg as? JsonObject)?.get("message")?.jsonPrimitive?.content
            throw BackendError(if (status == 429) "Too many reports for now. Try again in a minute." else detail ?: "The server refused the report.")
        }
    }

    suspend fun confirm(alertId: String, vote: String, token: String) {
        val (status, _) = Backend.send("POST", "/api/flare/confirm", token, """{"alert_id":${JsonPrimitive(alertId)},"vote":"$vote"}""")
        if (status != 200) throw BackendError("Could not record that.")
    }
}

/** The Waze-style report sheet: one tap on a kind, an optional note, send. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ReportSheet(model: DriveViewModel, lat: Double, lon: Double, onClose: () -> Unit) {
    var kind by remember { mutableStateOf<String?>(null) }
    var note by remember { mutableStateOf("") }
    var sending by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    val user by model.account.user.collectAsStateWithLifecycle()
    val busy by model.account.busy.collectAsStateWithLifecycle()
    val scope = rememberCoroutineScope()
    val context = LocalContext.current

    ModalBottomSheet(onDismissRequest = onClose, modifier = Modifier.testTag("report-sheet")) {
        Column(Modifier.padding(horizontal = 20.dp).padding(bottom = 32.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
            Text("Report", style = MaterialTheme.typography.titleLarge)
            if (user == null) {
                Row(Modifier.fillMaxWidth().background(MaterialTheme.colorScheme.surfaceVariant, RoundedCornerShape(12.dp)).padding(12.dp),
                    verticalAlignment = Alignment.CenterVertically) {
                    Text("Reports need an account, the same one as the website.", Modifier.weight(1f), style = MaterialTheme.typography.bodySmall)
                    TextButton({ scope.launch { (context as? Activity)?.let { model.account.signInWithGoogle(it) } } }, enabled = !busy) { Text("Sign in") }
                }
            }
            LazyVerticalGrid(GridCells.Adaptive(100.dp), Modifier.height(330.dp), verticalArrangement = Arrangement.spacedBy(10.dp), horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                items(ReportKinds.all, key = { it.kind }) { k ->
                    val on = kind == k.kind
                    Column(
                        Modifier.fillMaxWidth().height(84.dp)
                            .background(if (on) MaterialTheme.colorScheme.primary.copy(alpha = 0.18f) else MaterialTheme.colorScheme.surfaceVariant, RoundedCornerShape(12.dp))
                            .border(2.dp, if (on) MaterialTheme.colorScheme.primary else androidx.compose.ui.graphics.Color.Transparent, RoundedCornerShape(12.dp))
                            .clickable { kind = k.kind }.padding(8.dp).testTag("report-${k.kind}"),
                        horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.Center,
                    ) {
                        Icon(k.icon, null)
                        Spacer(Modifier.height(6.dp))
                        Text(k.label, style = MaterialTheme.typography.labelSmall, textAlign = TextAlign.Center, maxLines = 2)
                    }
                }
            }
            OutlinedTextField(note, { note = it }, Modifier.fillMaxWidth(), placeholder = { Text("Add a note (optional)") }, maxLines = 3)
            Text("Reported at %.5f, %.5f. Reports show on the map for everyone and go to the plugins you use, under a pseudonym.".format(lat, lon),
                style = MaterialTheme.typography.bodySmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            error?.let { Text(it, color = MaterialTheme.colorScheme.error, style = MaterialTheme.typography.bodySmall) }
            Button(
                onClick = {
                    val k = kind ?: return@Button
                    scope.launch {
                        val token = model.account.token()
                        if (token == null) { (context as? Activity)?.let { model.account.signInWithGoogle(it) }; return@launch }
                        sending = true
                        try {
                            Reporter.send(k, lat, lon, model.courseDegrees, note, token)
                            model.markers.refresh(true)
                            model.toast("Thanks. Your report is on the map.")
                            onClose()
                        } catch (e: Exception) {
                            error = e.message
                        } finally { sending = false }
                    }
                },
                enabled = kind != null && !sending, modifier = Modifier.fillMaxWidth().testTag("report-send"),
            ) { Icon(Icons.Default.Send, null); Spacer(Modifier.width(6.dp)); Text(if (sending) "Sending" else "Send report") }
        }
    }
}

File: ios/CommuteScoutDrive/Report.swift
import CoreLocation
import SwiftUI

/// What a driver can report, the website's list, in the order people
/// reach for them. Kinds are the Flare vocabulary.
enum ReportKinds {
    static let all: [(kind: String, label: String, symbol: String)] = [
        ("POLICE_VISIBLE", "Police", "shield.lefthalf.filled"),
        ("CRASH_MAJOR", "Crash", "car.side.rear.and.collision.and.car.side.front"),
        ("HAZARD_ON_ROAD", "Hazard on road", "exclamationmark.triangle.fill"),
        ("HAZARD_SHOULDER_CAR", "Car on shoulder", "car.side"),
        ("ROAD_CLOSED", "Road closed", "xmark.octagon.fill"),
        ("LANE_CLOSED", "Lane closed", "road.lanes"),
        ("JAM_HEAVY", "Traffic jam", "car.2.fill"),
        ("WEATHER_FOG", "Fog or weather", "cloud.fog.fill"),
        ("WEATHER_FLOOD", "Flooding", "water.waves"),
        ("WEATHER_ICE", "Ice or snow", "snowflake"),
        ("CHAINS_REQUIRED", "Chains required", "link"),
        ("CHAINS_NOT_REQUIRED", "Chains not needed", "link.badge.plus"),
        ("CAMERA_ISSUE", "Camera issue", "camera.fill"),
        ("MAP_ISSUE", "Map issue", "map.fill"),
    ]
}

/// Sends a report to commutescout.com, which fans it out to plugins.
@MainActor
final class Reporter: ObservableObject {
    @Published var sending = false
    @Published var lastSent: String?

    func send(kind: String, at coordinate: CLLocationCoordinate2D, heading: Double?, description: String,
              token: String) async throws {
        sending = true
        defer { sending = false }
        var body: [String: Any] = ["kind": kind, "lat": coordinate.latitude, "lon": coordinate.longitude,
                                   "client": "commutescout-ios/\(AppInfo.version)"]
        if let heading, heading >= 0 { body["heading_deg"] = heading }
        if !description.isEmpty { body["description"] = description }
        let (status, data) = try await Backend.send("POST", "api/flare/report", token: token, body: body)
        guard status == 201 || status == 202 else {
            let msg = (try? JSONSerialization.jsonObject(with: data) as? [String: Any])?["error"]
            throw ReportError.refused(status, (msg as? String) ?? (msg as? [String: Any])?["message"] as? String ?? "The server refused the report.")
        }
        lastSent = kind
    }

    func confirm(alertId: String, vote: String, token: String) async throws {
        let (status, _) = try await Backend.send("POST", "api/flare/confirm", token: token,
                                                 body: ["alert_id": alertId, "vote": vote])
        guard status == 200 else { throw ReportError.refused(status, "Could not record that.") }
    }
}

enum ReportError: LocalizedError {
    case refused(Int, String)
    var errorDescription: String? {
        switch self {
        case let .refused(code, msg): code == 429 ? "Too many reports for now. Try again in a minute." : msg
        }
    }
}

/// The Waze-style report sheet: one tap on a kind, an optional note, send.
/// The location is where the driver is (or the pin, when browsing one).
struct ReportSheet: View {
    @EnvironmentObject var model: AppModel
    @Environment(\.dismiss) private var dismiss
    let coordinate: CLLocationCoordinate2D
    @State private var kind: String?
    @State private var note = ""
    @State private var error: String?
    @State private var showSignIn = false

    private let columns = [GridItem(.adaptive(minimum: 96), spacing: 10)]

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 14) {
                    if !model.account.signedIn {
                        HStack {
                            Image(systemName: "person.crop.circle.badge.exclamationmark").foregroundStyle(.orange)
                            Text("Reports need an account, the same one as the website.").font(.footnote)
                            Spacer()
                            Button("Sign in") { showSignIn = true }.buttonStyle(.bordered).controlSize(.small)
                        }
                        .padding(12).background(.thinMaterial, in: RoundedRectangle(cornerRadius: 12))
                    }
                    LazyVGrid(columns: columns, spacing: 10) {
                        ForEach(ReportKinds.all, id: \.kind) { k in
                            Button { kind = k.kind } label: {
                                VStack(spacing: 6) {
                                    Image(systemName: k.symbol).font(.title2)
                                    Text(k.label).font(.caption).multilineTextAlignment(.center).lineLimit(2)
                                }
                                .frame(maxWidth: .infinity, minHeight: 74)
                                .padding(8)
                                .background(kind == k.kind ? Color.accentColor.opacity(0.18) : Color(.secondarySystemBackground),
                                            in: RoundedRectangle(cornerRadius: 12))
                                .overlay(RoundedRectangle(cornerRadius: 12).stroke(kind == k.kind ? Color.accentColor : .clear, lineWidth: 2))
                            }
                            .buttonStyle(.plain)
                            .accessibilityIdentifier("report-\(k.kind)")
                        }
                    }
                    TextField("Add a note (optional)", text: $note, axis: .vertical).lineLimit(1 ... 3)
                        .textFieldStyle(.roundedBorder)
                    Text("Reported at \(coordinate.pretty). Reports show on the map for everyone and go to the plugins you use, under a pseudonym.")
                        .font(.footnote).foregroundStyle(.secondary)
                    Button {
                        Task { await send() }
                    } label: {
                        Label(model.reporter.sending ? "Sending" : "Send report", systemImage: "paperplane.fill")
                            .frame(maxWidth: .infinity).padding(.vertical, 10)
                    }
                    .buttonStyle(.borderedProminent)
                    .disabled(kind == nil || model.reporter.sending)
                    .accessibilityIdentifier("report-send")
                }
                .padding(16)
            }
            .navigationTitle("Report")
            .toolbar { Button("Cancel") { dismiss() } }
            .sheet(isPresented: $showSignIn) { SignInSheet(reason: "Sign in to send reports.") }
            .alert("Could not send", isPresented: Binding(get: { error != nil }, set: { if !$0 { error = nil } })) {
                Button("OK") { error = nil }
            } message: { Text(error ?? "") }
        }
    }

    private func send() async {
        guard let kind else { return }
        guard let token = await model.account.token() else { showSignIn = true; return }
        do {
            try await model.reporter.send(kind: kind, at: coordinate, heading: model.courseDegrees, description: note, token: token)
            model.markers.refresh(force: true)
            model.toast = "Thanks. Your report is on the map."
            dismiss()
        } catch {
            self.error = error.localizedDescription
        }
    }
}
